iceoryx2-bb/cxx/include/iox2/container/static_vector.hpp

```diff
@@ -0,0 +1,262 @@
+    using OptionalConstReference = Optional<std::reference_wrapper<T const>>;
+
+    // Unchecked element access
+    class UncheckedConstAccessor {
+        friend class StaticVector;
+    private:
+        StaticVector const* m_parent;
+    private:
+        constexpr UncheckedConstAccessor(StaticVector const& parent)
+        :m_parent(&parent)
+        {}
+    public:
+        UncheckedConstAccessor& operator=(UncheckedConstAccessor&&) = delete;
+
+        constexpr ConstReference operator[](SizeType index) const {
+            return *m_parent->m_storage.pointer_from_index(index);
+        }
+
+        constexpr ConstIterator begin() const noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr ConstIterator end() const noexcept {
+            return m_parent->m_storage.pointer_from_index(m_parent->m_storage.size);
+        }
+
+        constexpr ConstPointer data() const noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+    };
+
+    class UncheckedAccessor {
+        friend class StaticVector;
+    private:
+        StaticVector* m_parent;
+    private:
+        constexpr UncheckedAccessor(StaticVector& parent)
+        :m_parent(&parent)
+        {}
+    public:
+        UncheckedAccessor& operator=(UncheckedAccessor&&) = delete;
+
+        constexpr Reference operator[](SizeType index) {
+            return *m_parent->m_storage.pointer_from_index(index);
+        }
+
+        constexpr ConstReference operator[](SizeType index) const {
+            return *m_parent->m_storage.pointer_from_index(index);
+        }
+
+        constexpr Iterator begin() noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr ConstIterator begin() const noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr Iterator end() noexcept {
+            return m_parent->m_storage.pointer_from_index(this->m_parent->m_storage.size);
+        }
+        
+        constexpr ConstIterator end() const noexcept {
+            return m_parent->m_storage.pointer_from_index(this->m_parent->m_storage.size);
+        }
+
+        constexpr Pointer data() noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+
+        constexpr ConstPointer data() const noexcept {
+            return m_parent->m_storage.pointer_from_index(0);
+        }
+    };
+private:
+    template<uint64_t M> friend class StaticVector<T, M>;
+    detail::StaticVectorStorage<T, N> m_storage;
+public:
+    // constructors
+    constexpr StaticVector() noexcept = default;
+    constexpr StaticVector(StaticVector const&) = default;
+    constexpr StaticVector(StaticVector&&) = default;
+    
+    template<uint64_t M, typename std::enable_if<(N >= M), bool>::type = true>
+    constexpr StaticVector(StaticVector<T, M> const& rhs)
+    :m_storage(rhs.m_storage)
+    {}
```

Comment:
The idea here is that the capacity is not a salient part of the container's value, so it should be possible for vectors of different capacities to interact. This will also expand to things like comparison operators.

Reply:
Sure, that sounds like a useful facility.